Machine learning (#27)
bcgov/ds-intro-to-python: course-content/34_machine_learning.qmd

While there are other libraries outside of sklearn that support machine learning in Python (e.g. PyTorch, Tensorflow), it is widely used in the data science community and is known for its ease of use and efficient implementation of algorithms. 

Now that you know some of the basics, the best way to see how machine learning is done is through tackling a meaningful problem with some real data. 

## Defining a Problem

Imagine you are a server at a restaurant and want to maximize your tip-earning potential. Being able to predict which customers are more likely to leave you a big tip would definitely be useful!  This is a nice, if simple, use case that machine learning can help us with.  

<center>
    <img src="images/machine-learning/tipping.PNG" width=90% style="margin:auto" border=1/>
    <p style="text-align: center">
    Decisions, decisions!
    </p>
</center>

> It is also a classic "supervised" machine learning use case, where we will need to have both input and output data available to analyze. By supervised, we mean that in our data we have the target data that we want to be able to predict (see workflow for a supervised use case in the image below). You can also think of the workflow as a high level pipeline. 

Reviewer: ```suggestion
It is also a classic "supervised" machine learning use case, where we will need to have both input and output data available to analyze. By **supervised**, we mean that in our data we have the target data, or outputs, that we want to be able to predict (see workflow for a supervised use case in the image below). You can also think of the workflow as a high level pipeline. 
```

<center>
    <img src="images/machine-learning/supervised_learning_workflow.PNG" width=70% style="margin:auto" border=1/>
    <p style="text-align: center">
        Supervised machine learning workflow
    </p>
</center>

## Getting necessary Data

We find out that there is a small, but thankfully clean and available dataset, made available by Seaborn. It contains restaurant visit data from 244 parties of customers to a given restaurant. Each row of data represents a restaurant visit by one party of one or more people. Below are the fields included in the dataset:

- tip in dollars <-- our target variable
- bill in dollars
- sex of the bill payer
- whether there were smokers in the party
- day of the week
- time of day
- size of the party

 We already know that there is no missing data, so that's great.  Let's run the `sample()` method on the dataframe to get a good balanced initial feel for the data. 

```{python}
# importing libraries for getting the data and pre-processing it

> import seaborn as sb

Reviewer: ```suggestion
import seaborn as sns
```

Seaborn always imported as sns

import pandas as pd

> tips_df = sb.load_dataset('tips')

Reviewer: ```suggestion
tips_df = sns.load_dataset('tips')
```

tips_df.sample(10)
```

## Choosing between Regression and Classification 

> With our tip predictor project, there are two approaches we could take and we must decide which one to take at the outset. One approach is to predict the *actual size of the tip* one would expect to get. This is what would be known as a regression type problem. The other way is to define *what constitutes a bigger tip* from the outset and to predict how likely you would be to get that. This is called a classification type of problem. There is no correct answer, is use-case dependent, and each have their pros and cons. 

Reviewer: ```suggestion
With our tip predictor project, there are two approaches we could take and we must decide which one to take at the outset. One approach is to predict the *actual size of the tip* one would expect to get. This is what would be known as a regression type problem. The other way is to define *what constitutes a bigger tip* from the outset and to predict how likely you would be to get that. This is called a classification type of problem. There is no correct answer: the model used is use-case dependent, and each will have their own pros and cons. 
```

To accomplish this, let's run some code using `lambda` and `apply()` to compute a big tipper variable and then take another look at the dataframe. 

```{python}
tips_df['big_tipper'] = tips_df['tip'].apply(lambda x: 0 if x < 2.9 else 1)
tips_df.sample(5)
```

Well that seemed to work fine! The machine learning classification model that we will use can only use numeric data. You probably know from statistics that you can convert categorical variables into numeric ones by turning each of the subsets into discrete columns consisting of 1s and 0s. They are widely referred to as dummy variables. Dummies are like flags in the data, where 1 means a certain subset is "observed" and 0 means that particular subset is "absent". 

Fortunately, there is a method called `get_dummies()` in pandas that can transform all non-numeric variables easily. In the code below, we pass in the whole dataframe and it will do its thing.

```{python}
tips_df = pd.get_dummies(tips_df)
tips_df.head()
```
<br>

> At this point, we could go ahead and eliminate some of these newly created columns. For example, we could drop "time_Lunch" as this is essentially computationally redundant to "time_Dinner", just the exact obverse (i.e. if you know whether a meal was lunch or not, you know whether it was dinner or not) But we'll leave that for now, their continued presence in the data will not harm the model.

Reviewer: Isn't there an issue with leaving in perfectly correlated variables? At least when we want to try to learn about the impacts of different coefficients.

Independent variables are usually best thought of as those which are known at the point of decision. In our use case, we want to predict whether a party is likely to be a big tipper when they arrive at the restauarant.  At that point we likely know the number of people in the party, but not the total bill amount. So let's identify the independent variables or features by dropping the variables 'tip', 'total_bill', and 'big_tipper' from an object we will call "X". The dependent or target variable is easier to identify. That is simply "big_tipper". We will assign that a "y". Both "X" (typed usually as upper case X) and "y" (typed usually as lower case y) are widely used conventions in Python for these concepts. 

Let's use what we learned about the `drop()` method to create the features object and assign to y the single column containing the target "big_tipper" variable.

```{python}
X = tips_df.drop(['tip', 'total_bill', 'big_tipper'], axis=1)
y = tips_df['big_tipper']
```

> One of the key requirements of supervised machine learning methods is to split the dataset up into training and test samples. For the training set, the "machine" will "learn" which restaurant parties are more likely to be big tippers and use that knowledge to create an algorithm that can be applied to the test set. 

Reviewer: I'd probably include a reason, or a link to an article, as to why we want training/test samples. I assume that none of this will be common knowledge for anyone.

Author: Fair!

Scikit-Learn makes it easy with its famous `train_test_split()` method:

```{python}
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=155, test_size=0.409)
```

> The `train_test_split()` method takes the X and y that we defined above, and splits it randomly according to the relative size we want the test sample to be compared to the total sample size. In our example, we have chosen for the test sample to be 40.9% of the entire dataset.  The percentage size you will choose will depend on a number of factors relating to the use case. 

Reviewer: Explain the random state as well

Author: Will do.

```{python}
from sklearn.linear_model import LogisticRegression

reg = LogisticRegression()
reg.fit(X_train, y_train)
```
<br>
Now our we have a trained model and we can run that model on our test dataset. We enter X_test as an argument following the `predict()` and `predict_proba()` methods as in the code below. Let us look at the first five cases in our array of predictions:

```{python}
y_predict =  reg.predict(X_test) # makes a decision based on probability
y_predict[:5] # first 5 prediction decisions in test dataset
```
The output gives us an array of predictions for the first five cases in sequence in the test dataset. A value of 1 means that the model predicts the party to be a big tipper, and 0 predicts that it will not be a big tipper. 

Each of the predictions is based on a discrete probability. The `predict_proba()` method returns the class probabilities for each data point.  The number of probabilities for each row is equal to the number of categories in target variable (2 in our case).

```{python}
y_predict_proba =  reg.predict_proba(X_test) # computes a probability per case
y_predict_proba[:5] # first 5 predictions in test dataset
```

> We can see for each case a pair of probabilities, one for "0" and the other for "1". This detail helps us understand how the decisions with the `predict()` method were made. 

Reviewer: ```suggestion
We can see for each case a pair of probabilities, one for "0" and the other for "1". For example, for the first set of diners in our test dataset, they had a 48% probability of leaving a low tip, while a 52% probability of leaving a large tip. This detail helps us understand how the decisions with the `predict()` method were made. 
```

## Evaluating

> Ok. We have our predictions... but are they good? Of course, what "good" means needs a bit of thought. There are many performance measures in data science and there are considerations based on the use cases themselves. How we would evaluate performace of an algorithm that attempts to diagnose serious disease would be different than if we were evaluating one that identified whether an email was spam or not. 

Reviewer: ```suggestion
Ok. We have our predictions... but are they good? Of course, what "good" means needs a bit of thought. There are many performance measures in data science and there are considerations based on the use cases themselves. How we would evaluate performance of an algorithm that attempts to diagnose serious disease would be different than if we were evaluating one that identified whether an email was spam or not. In the case of disease classification, we would not want a high false negative rate, and thus missing out on important treatment for people who actually have the disease. On the other hand with spam filters, it would be okay to miss some spam, but it would be worse to have a high false positive rate, and start chucking out real emails with possible spam. 
```

Author: I reworked this content into the discussion of precision and recall

Another common performance measuring tool is the so-called "confusion matrix". 

```{python}
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, y_predict)) 
```

<center>
    <img src="images/machine-learning/confusion_matrix.PNG" width=70% style="margin:auto" border=1/>
    <p style="text-align: center">
    Four quadrants of the confusion matrix
    </p>
</center>

The confusion matrix shows how many guesses the model gets "right", that is, the actual is true when we guess true (TP - True Positive) and false when we guess false (TN - True Negative). The remaining two quadrants show the errors. Type I error is the false positive error (FP - False Positive), where the model guesses true but reality is false. Type II error is where where the model guesses false but the reality is true (FN - False Negative).
<br>
<br>
Another widely used report is the classification report, which shows precision and recall, which are two widely reported metrics in machine learning. 

```{python}
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))
```

> **Precision** = True Positives / (True Positives + False Positives) <br>

Reviewer: I find these are easier to wrap my head around with a little example/ english definition, but up to you. I like the cast a wide/narrow net analogy.

Author: Good idea. Net analogy is good.  Don't want to wade too deeply into these measures, as its more about the coding, but point very well taken. These are easy terms to forget from point 1 to point 2, as they are laden with cognitive baggage from other contexts.

Reviewer: Yeah I agree, keep it simple!
